src/simplifiers/LARewriter.cc

```diff
@@ -0,0 +1,111 @@
+//
+// Created by Martin Blicha on 03.02.21.
+//
+
+#include "LARewriter.h"
+
+#include "LIALogic.h"
+#include "OsmtApiException.h"
+
+std::size_t DivModRewriter::counter = 0;
+const char * DivModRewriter::divPrefix = ".div";
+const char * DivModRewriter::modPrefix = ".mod";
+
+DivModRewriter::DivModPair DivModRewriter::freshDivModPair(LIALogic & logic) {
+    std::string num = std::to_string(counter++);
+    std::string divName = divPrefix + num;
+    std::string modName = modPrefix + num;
+    return {logic.mkNumVar(divName.c_str()), logic.mkNumVar(modName.c_str())};
+}
+
+PTRef rewriteDivMod(LIALogic & logic, PTRef root) {
+    return DivModRewriter().rewriteDivMod(logic, root);
+}
+
+// TODO: figure out how to cache assigned variables to avoid duplication
+PTRef DivModRewriter::rewriteDivMod(LIALogic &logic, PTRef root) {
+    if (root == PTRef_Undef or not logic.hasSortBool(root)) {
+        throw OsmtApiException("Div/Mod rewriting should only be called on formulas, not terms!");
+    }
+    vec<PTRef> definitions;
+    struct DFSEntry {
+        DFSEntry(PTRef term) : term(term) {}
+        PTRef term;
+        unsigned int nextChild = 0;
+    };
+    // MB: Relies on an invariant that id of a child is lower than id of a parent.
+    auto size = Idx(logic.getPterm(root).getId()) + 1;
+    std::vector<char> done;
+    done.resize(size, 0);
+    std::unordered_map<PTRef, PTRef, PTRefHash> substitutions;
+    std::vector<DFSEntry> toProcess;
+    toProcess.push_back(DFSEntry{root});
+    while (not toProcess.empty()) {
+        auto & currentEntry = toProcess.back();
+        PTRef currentRef = currentEntry.term;
+        Pterm const & term = logic.getPterm(currentRef);
+        unsigned childrenCount = term.size();
+        if (currentEntry.nextChild < childrenCount) {
+            PTRef nextChild = term[currentEntry.nextChild];
+            ++currentEntry.nextChild;
+            if (done[Idx(logic.getPterm(nextChild).getId())] == 0) {
+                toProcess.push_back(DFSEntry(nextChild));
+            }
+            continue;
+        }
```

Comment:
I've been writing the above code slightly different, see for instance [the corresponding loop in IteToSwitch.cc](https://github.com/usi-verification-and-security/opensmt/blob/e33b1470696e02e9360f4091ceadfdf8b709657d/src/itehandler/IteToSwitch.cc#L40).  Not sure which one is better, but I guess we should settle for one version.  I think grepping `unprocessed_children` would reveal most of these cases.

Reply:
This is probably an argument for introducing the general visit pattern.  Or maybe that's yet another topic since we don't do the rewriting in all of the cases.

Reply:
I am thinking that we could have a const visitor that does not modify the terms, just collects information and than separately a rewriter that rewrites the terms.